api/src/services/specifications.ts

```diff
@@ -579,3 +582,120 @@ class GraphQLSpecsService implements SpecificationSubService {
 		return null;
 	}
 }
+
+class TypescriptSpecsService implements SpecificationSubService {
+	accountability: Accountability | null;
+	knex: Knex;
+	schema: SchemaOverview;
+
+	fieldsService: FieldsService;
+
+	constructor(
+		options: AbstractServiceOptions,
+		{
+			fieldsService,
+		}: {
+			fieldsService: FieldsService;
+		}
+	) {
+		this.accountability = options.accountability || null;
+		this.knex = options.knex || getDatabase();
+		this.schema = options.schema;
+
+		this.fieldsService = fieldsService;
+	}
+
+	async generate() {
+		const collections = await this.getCollections();
+		return (
+			(await Promise.all(collections.map(async (collection) => await this.generateCollectionType(collection)))).join(
+				'\n\n'
+			) +
+			'\n\n' +
+			this.generateDefaultExport(collections)
+		);
+	}
+
+	async getCollections() {
+		return (
+			await Promise.all(
+				Object.values(this.schema.collections).map(async (collection) => {
+					const fields = (await this.fieldsService.readAll(collection.collection)).filter(
+						(field) => !field.meta?.system
+					);
+					return {
+						name: collection.collection,
+						pascalName: this.tryGetPascalName(collection.collection, Object.keys(this.schema.collections)),
+						fields,
+					};
+				})
+			)
+		).filter((collection) => collection.fields.length > 0);
+	}
+
+	async generateCollectionType(collection: { name: string; pascalName: string; fields: Field[] }) {
+		let collectionType = `export type ${collection.pascalName} = {\n`;
+		collectionType += Object.values(collection.fields)
+			.map((field) => `  ${field.field}${field.schema?.is_nullable ? '?' : ''}: ${this.getType(field.type)};`)
```

Comment:
This should take into account relations to be usable with the sdk's `fields` option. Right now the types barf at something like `fields: ['image.id']`.

In my local version of this I use this:
![image](https://user-images.githubusercontent.com/9959940/194773380-948a4c58-ee92-4e18-89e9-8e7c1c10ede4.png)

Reply:
Well, in the meantime I adjusted my code after reading https://github.com/directus/sdk/discussions/34#discussioncomment-3359536 and to include m2m relations:

```ts
function getRawType(directusType: string): string {
  if (['integer', 'bigInteger', 'float', 'decimal'].includes(directusType)) return 'number';
  if (['boolean'].includes(directusType)) return 'boolean';
  if (['json'].includes(directusType)) return 'any[] | Record<string, any>';
  if (['csv'].includes(directusType)) return 'any[]';

  return 'string';
}

function getType(collection: string, field: Field, relations: Relation[]): string {
  const joinRelation = relations.find(
    (rel) => rel.related_collection === collection && rel.meta?.one_field === field.field,
  );
  if (joinRelation) {
    return `${getRawType(field.type)}[] | ${joinRelation.collection}[]`;
  }

  const relation = relations.find((rel) => rel.collection === collection && rel.field === field.field);
  if (relation?.related_collection) {
    return `${getRawType(field.type)} | ${relation.related_collection}`;
  }

  return getRawType(field.type);
}
```

My next step is to exclude types for relations the user has no access to.